### PR TITLE
WIP: create base class for builders

### PR DESCRIPTION
**Pull request by CJ-Wright**
The basic idea is that all builders have a build method, which
run all the methods specified in `self.cmds`. Eventually these
commands may be pulled from the rc (eg some people want to build the
people page but not the project page). This will allow the number of
actions in a builder to rise without users having to run all of them.

### CODE REVIEW

File: regolith/builder_base.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+"""Builder Base Class"""
+import os
+
+from jinja2 import Environment, FileSystemLoader
+
+
+class BuilderBase(object):
```
Comment (scopatz) [inline regolith/builder_base.py]: Docstrings needed throughout this file

File: regolith/cvbuilder.py [hunk 2]
```diff
@@ -1,47 +1,35 @@
 """Builder for CVs."""
 import os
-import shutil
 import subprocess
 from glob import glob
-from copy import deepcopy
 from itertools import groupby
 
-from jinja2 import Environment, FileSystemLoader
+from regolith.builder_base import BuilderBase
```
Comment (scopatz) [inline regolith/cvbuilder.py]: This import should be moved to be down with the other regolith imports

File: regolith/cvbuilder.py [hunk 3]
```diff
@@ -1,47 +1,35 @@
...
-        self.construct_global_ctx()
-        if HAVE_BIBTEX_PARSER:
-            self.bibdb = BibDatabase()
-            self.bibwriter = BibTexWriter()
+        super().__init__(rc)
+        self.cmds = ['latex', 'pdf', 'clean']
 
     def construct_global_ctx(self):
-        self.gtx = gtx = {}
+        gtx = self.gtx
```
Comment (scopatz) [inline regolith/cvbuilder.py]: Some of these basic global context constructions (len, true, false, etc) can be moved to the base class

**Comment by scopatz** [comment]
This generally looks like the right idea but needs docs etc.

**Comment by CJ-Wright** [comment]
woo still net negative!

File: regolith/builder_base.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        gtx['category_val'] = category_val
+        gtx['rfc822now'] = rfc822now
+        gtx['date_to_rfc822'] = date_to_rfc822
+
+    def render(self, tname, fname, **kwargs):
+        """Render the template into a file using the kwargs and global context
+
+        Parameters
+        ----------
+        tname: str
```
Comment (scopatz) [inline regolith/builder_base.py]: numpydoc requires space on both sides of the colon

File: regolith/cvbuilder.py [hunk 5]
```diff
@@ -1,100 +1,52 @@
...
-    level_val, id_key, date_key, position_key
-
-LATEX_OPTS = ['-halt-on-error', '-file-line-error']
-
-def latex_safe(s):
-    return s.replace('&', '\&').replace('$', '\$').replace('#', '\#')
-
-class CVBuilder(object):
 
+class CVBuilder(BuilderBase):
```
Comment (scopatz) [inline regolith/cvbuilder.py]: Needs docs

File: regolith/cvbuilder.py [hunk 6]
```diff
@@ -1,100 +1,52 @@
...
-                    os.path.join(os.path.dirname(__file__), 'templates'),
-                    ]))
-        self.construct_global_ctx()
-        if HAVE_BIBTEX_PARSER:
-            self.bibdb = BibDatabase()
-            self.bibwriter = BibTexWriter()
+        super().__init__(rc)
+        self.cmds = ['latex', 'pdf', 'clean']
 
     def construct_global_ctx(self):
```
Comment (scopatz) [inline regolith/cvbuilder.py]: needs docs

File: regolith/cvbuilder.py [hunk 7]
```diff
@@ -1,100 +1,52 @@
...
-        with open(os.path.join(self.bldir, fname), 'wt') as f:
-            f.write(result)
-
-    def build(self):
-        os.makedirs(self.bldir, exist_ok=True)
-        self.latex()
-        self.pdf()
-        self.clean()
-
     def latex(self):
```
Comment (scopatz) [inline regolith/cvbuilder.py]: needs docs

File: regolith/htmlbuilder.py [hunk 8]
```diff
@@ -1,84 +1,34 @@
...
 
-from regolith.tools import all_docs_from_collection, date_to_float, date_to_rfc822, \
-    rfc822now, gets
-from regolith.sorters import doc_date_key, ene_date_key, category_val, \
-    level_val, id_key, date_key, position_key
-
-
-class HtmlBuilder(object):
 
+class HtmlBuilder(BuilderBase):
```
Comment (scopatz) [inline regolith/htmlbuilder.py]: docstring

File: regolith/tools.py [hunk 9]
```diff
@@ -142,3 +101,117 @@ def month_and_year(m=None, y=None):
     m = month_to_int(m)
     return '{0} {1}'.format(SHORT_MONTH_NAMES[m], y)
 
+
+def filter_publications(citations, authors, reverse=False, bold=True):
```
Comment (scopatz) [inline regolith/tools.py]: Docs here too...

File: regolith/dates.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
+MONTHS = {
```
Comment (scopatz) [inline regolith/dates.py]: This module needs a docstring

File: docs/api/index.rst [hunk 11]
```diff
@@ -27,3 +27,4 @@ For those of you who want the gritty details.
     commands
     main
     app
+    interactive
```
Comment (scopatz) [inline docs/api/index.rst]: shouldn't dates and the other new files be added here?

Comment (CJ-Wright) [reply docs/api/index.rst]: Is there a nice command for that or should I write everything by hand?

Comment (scopatz) [reply docs/api/index.rst]: I think you need to do it by hand?

**Comment by scopatz** [comment]
Thanks @CJ-Wright!

**Pull request merged**